src/cirrus/management/deployment.py

```
@@ -37,68 +41,52 @@ def _maybe_use_buffer(fileobj: IO):
-
-    def save(self, path: Path) -> int:
-        return path.write_text(self.asjson(indent=4))
-
-    def asdict(self) -> dict[str, Any]:
-        return dataclasses.asdict(self)
-
-    def asjson(self, *args, **kwargs) -> str:
-        return json.dumps(self.asdict(), *args, **kwargs)
-
-
-@dataclasses.dataclass
-class Deployment(DeploymentMeta):
+class Deployment:
     def __init__(
         self,
-        *args,
+        name: str,
+        environment: dict,
         session: boto3.Session | None = None,
-        **kwargs,
     ) -> None:
-        super().__init__(*args, **kwargs)
+        self.name = name
+        self.environment = environment
         self.session = session if session else boto3.Session()
         self._functions: list[str] | None = None
 
     @staticmethod
     def yield_deployments(
-        region: str | None = None,
-        session: boto3.Session | None = None,
-    ) -> Iterator[DeploymentPointer]:
-        yield from DeploymentPointer.list(region=region, session=session)
+        session: boto3.Session,
+    ) -> Iterator[str]:
+        yield from DeploymentPointer.list_deployments(session=session)
 
     @classmethod
     def from_pointer(
         cls,
         pointer: DeploymentPointer,
-        session: boto3.Session | None = None,
-    ) -> Deployment:
-        return cls(session=session, **pointer.get_config(session=session))
+        session: boto3.Session,
+    ):
+        return cls(session=session, **pointer.get_configuration(session=session))
 
     @classmethod
-    def from_name(cls, name: str, session: boto3.Session | None = None) -> Deployment:
-        dp = DeploymentPointer.get(name, session=session)
-        return cls.from_pointer(dp, session=session)
+    def from_name(cls, name: str, session: boto3.Session) -> Deployment:
+        try:
+            dp = DeploymentPointer.get_pointer(name, session=session)
+            return cls.from_pointer(dp, session=session)
+        except ClientError:
```

this gets raised if we can't find the pointer parameter in parameter store.  This seemed the appropriate place for this error because it keeps deployment_pointer agnostic if there is a pointer parameter or not, it just looks for it.  If it doesn't find it then we can't find a deployment and so then we can't build the `Deployment` class i.e. unable to get vars = unable to build deployment object for other work.

---

Hmm, well, I disagree about this being the appropriate place for this exception handling. 😅 Your reasoning sounds good, but I then interpret that reasoning to mean this should be implemented a bit differently.

The fact that a boto3 `ClientError` could be thrown is not obvious when calling `DeploymentPointer.get_pointer()`, in particular because it is not a boto3 function. doubly so that `ClientError` means what you are interpreting it to mean here, that the deployment was not found. Typically if you are calling a function from a third-party dependency and that can throw an exception you want to make sure you handle that exception at the call site to translate it into something more meaningful in that one place. As this code is written the implementation details of `DeploymentPointer.get_pointer()` are effectively leaking into this calling context.

Also, I'm not sure that handling `ClientError` indiscriminately like this is correct. `ClientError`, in my experience, is thrown for _any_ error the client encounters, not all of which would indicate the deployment doesn't exist (such as hitting a rate limit, network issues, etc.). Often proper handling of a `ClientError` requires further inspection of the error. [Here's an example](https://github.com/cirrus-geo/cirrus-geo/blob/main/src/cirrus/lib/process_payload.py#L244) from a different part of the cirrus codebase showing how only two specific `ClientErrors` could be handled in that context, and anything else has to be re-raised.

---

Interesting.  I get what you mean in regards to handling error from a 3rd party API call at the site of the API that makes sense instead of letting it float up and be handled elsewhere, possibly making it harder to track.   That's a helpful example to link to.

---

Moved error handling into `deployment_pointer.py` to wrap the API call that could/should raise the error if we can't find a deployment.